// geofeedia-lumberjack/config.js
'use strict';

// ** Dependencies
const path = require('path');
const lumberjack = require('./lumberjack')

/**
 * Loads the value of a configuration variable.
 * @param variable
 */
const $ = env_var => process.env[env_var];

/**
 * Ensure value conforms to ENV variable format specification.
 * - All uppercase
 * - Replace dash '-' with underscores '_'
 * @param val
 * @returns {string}
 */
const sanitize = val => val.toUpperCase().replace(/-/g, '_');

// ** Load from ENV var SERVICE_NAME and replace underscores with dashes
const service_name = $('SERVICE_NAME') || '';
const env_prefix = sanitize(service_name) + '_';

/**
 * Load an application specific configuration variable from the environment
 * @param name
 * @param default_value
 * @returns {jQuery|HTMLElement}
 */
const config = (name, default_value) => $(env_prefix + sanitize(name)) || default_value;

module.exports = {
    logging: {
        name: 'geofeedia.com.' + service_name,
        serviceName: service_name,
        useStdOut: true,
        fluent: {
            minLevel: config('LOG_LEVEL')
                ? lumberjack.LEVELS[config('LOG_LEVEL').toUpperCase()]
                : 30,
            // check for both env vars for backwards compat.
            host: process.env['FLUENTD_FORWARDER_V2_SERVICE_HOST'] || process.env['FLUENTD_V0_SERVICE_HOST'] || 'k8s-proxy.qa1.us-east-1.local', // for docker environment use: 'docker.local'
            port: process.env['FLUENTD_FORWARDER_V2_SERVICE_PORT'] || process.env['FLUENTD_V0_SERVICE_PORT'] || 31442 // for docker environment use: 24224
        }
    }
};
